### DISCUSSION TITLE
How to access hono routing params?

### DISCUSSION THREAD

**Discussion by FraBle** [magne4000/universal-middleware#27] [General]
Hono supports [named path parameters](https://hono.dev/docs/api/routing#path-parameter). How would one access those using the UniversalHandler?

Example

```typescript
app.get('/user/:name', (c) => {
  const name = c.req.param('name')
  ...
})
```

**Comment by magne4000** [comment]
I was going to say that named path parameters are not a universal concept, but in fact, many servers do implement this feature ([fastify](https://fastify.dev/docs/latest/Reference/Routes/#url-building), [express](https://expressjs.com/en/guide/routing.html#route-parameters), [hattip](https://www.npmjs.com/package/@hattip/router), [hono](https://hono.dev/docs/api/routing#path-parameter), [h3](https://h3.unjs.io/guide/router#route-params)).

For servers that support it, I could add route parameters under the [runtime argument](https://github.com/magne4000/universal-middleware/blob/616031d4536054958debd10d5c8b3f56fe20948d/packages/core/src/types.ts#L95) (which is not yet documented, as its still WIP).

**But what to do when an adapter does not support support it?** For instance the Cloudflare adapter as no concept of routes.

I guess that something like that could do the trick:
```ts
import { parseRoute } from `@universal-middleware/core`;

interface Options {
  route?: '/user/:name'
}

export const handler = ((options?: Options) => async (request, context, runtime) => {
  const name =
    // hypothetical route params on `runtime`
    runtime.params?.name ??
    // hypothetical `parseRoute` helper as a fallback, using `path-to-regexp` for instance
    parseRoute(request, options);
  
  if (!name) throw new Error("'name' route parameter required");
  
  // ...
}) satisfies Get<[Options | undefined], UniversalHandler>;
```

**Reply by magne4000** [reply]
It could be even more generic

```ts
interface Options {
  extractName?: (request: Request, context: Universal.Context, runtime: Runtime) => string;
}

export const handler = ((options?: Options) => async (request, context, runtime) => {
  const name =
    runtime.route('name') ??
    options?.extractName(request, context, runtime);
  
  if (!name) throw new Error("'name' route parameter required");
  
  // ...
}) satisfies Get<[Options | undefined], UniversalHandler>;
```

**Comment by magne4000** [comment]
Added by #29.

[Route parameters recipe](https://universal-middleware.dev/recipes/params-handler) example.

**Reply by FraBle** [reply]
Thank you for adding this feature and extensive example in such a short time. 🙇
